src/main/java/bsh/Interpreter.java

```diff
@@ -143,11 +143,12 @@ public static void setShutdownOnExit(final boolean value) {
 
     /* --- Instance data --- */
 
-    transient Parser parser;
+    protected transient Parser parser;
+    protected transient Reader in;
+    protected transient PrintStream out;
+    protected transient PrintStream err;
+
```

Comment:
The problem with doing this...and the reason I pointed out the that it is suggested to make all public classes final (so that you cannot extend them) is because we need to be extra careful not to change any public members as they are breaking. The same goes for making members protected, we now cannot change them in the future as we have to consider subclasses in use.

The package private is a good compromise as it doesn't completely disable the ability to extend the class or inherit its members but neither does it advertise it as "supported api"

Personally I think parser should be private, it can change the reliability of the language if we allow access to it. Also the in, out and err should be referenced from console instead IMO

Reply:
well, I may agree in principle (even if only up to a certain point), but to me this is mostly upfront thinking. do we really have a problem of tons of people creating subclasses of Interpreter? I would really love so, but may be it won't be the case for the next few months. In any case, please note that they are not private now. They have package visibility. This does not give you the benefits you are wishing, and makes subclassing odd (why foo project should have a bsh package to extend Interpreter?)